jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java

```diff
@@ -307,66 +345,220 @@ private void closeOperationHandle(TOperationHandle stmtHandle) throws SQLExcepti
+                result.fetchedRows = RowSetFactory.create(results, protocol);
+                if (useConvertedResultSet) {
+                  result.fetchedRows =
+                      new ConvertedResultSet(result.fetchedRows, getSchema());
+                }
+                result.startRow = results.getStartRowOffset() + 1;
+                if (hasStartRow && result.startRow < nextStartRow.get()) {
+                  throw new SQLException("Unexpected row offset");
+                }
+              } catch (Exception e) {
+                result.ex = e;
+              }
+
+              try {
+                // Wait for earlier row sets to be added to the queue
+                synchronized(nextStartRow) {
+                  if (!poolDone) {
+                    if (result.ex == null) {
+                      if (hasStartRow) {
+                        while (nextStartRow.get() != result.startRow) {
+                          nextStartRow.wait();
+                        }
+                        nextStartRow.set(result.startRow + result.numRows);
+                      }
+                      poolDone = !result.hasMoreRows;
+                    } else {
+                      poolDone = true;
+                    }
+                    resultQueue.put(result);
+                    if (hasStartRow) {
+                      nextStartRow.notifyAll();
+                    }
+                  }
+                }
+              } catch (InterruptedException e) {
+                interruptException = e;
+                break;
+              }
+            }
+            // TODO: shutdown called here can make execute return
+            // RejectedExecutionException
+            // pool.shutdownNow();
+          });
+        }
+      }
+      try {
+        if (interruptException != null) {
+          throw interruptException;
+        }
+        FetchResult result = resultQueue.take();
+        fetchDone = !result.hasMoreRows;
+        if (result.ex != null) {
+          ShutdownPool();
+          throw new SQLException(result.ex);
+        }
+        fetchedRows = result.fetchedRows;
+      } catch (InterruptedException e) {
+        throw new SQLException(e);
       }
+    }
 
-      if (!fetchedRowsItr.hasNext()) {
-        return false;
+    orientation = TFetchOrientation.FETCH_NEXT;
+    fetchedRowsItr = fetchedRows.iterator();
+
+    return fetchedRowsItr.hasNext();
+  }
+
+  void DrainQueue() throws SQLException {
+    FetchResult result;
+    while ((result = resultQueue.poll()) != null) {
+      if (result.ex != null) {
+        throw new SQLException(result.ex);
       }
+    }
+  }
 
-      row = fetchedRowsItr.next();
-      rowsFetched++;
-    } catch (SQLException eS) {
-      throw eS;
-    } catch (Exception ex) {
-      throw new SQLException("Error retrieving next row", ex);
+  void ShutdownPool() throws SQLException {
+    if (pool != null) {
+      poolDone = true;
+      DrainQueue();
```

Comment:
do we need to call it before pool shutdown?

Reply:
This sequence is necessary to ensure that nothing blocks and nothing is left in the queue.